README
  Yet Another GUI framework for Erlang - gtknode

  DESIGN GOALS

* GUI separated from application by message passing
* distributed (application and GUI can run on different machines)
* centered around a GUI builder
* small volume of (hand-written) code
* pre-documented

  ARCHITECTURE

  a c-node (a daemon that implements the Erlang distribution protocol)
that instantiates the GUI from a configuration file. the c-node sends
messages to the Erlang node when the user interacts with the GUI; the
Erlang application changes the state of the GUI by sending messages to
widgets in the c-node. the widgets should look like Erlang processes
with registered names. the protocol should look something like this.

CnodePid ! {aWidget,dosomething,Args}       % erlang->cNode
ApplicationPid ! {reply, Val}               % cNode->Erlang
ApplicationPid ! {signal,aWidget,eventType} % cNode->erlang

  in this example aWidget is the name given to a widget in the
configration file. it can also be thought of as the registered name of
the process implementing the widget in the c-node.
  the c-node is responsible for garbage-collecting all temporary data.

  IMPLEMENTATION

  i chose GTK2 (www.gtk.org) as the framework. there were several reasons;

* it has an eminent GUI builder, Glade (glade.gnome.org)
* it has facilities for instantiating the GUI from the Glade (XML) files
* it has good documentation
* it supports run-time type checking
* its object orientation maps well to Erlang (Obj.meth(Arg) -> Obj!{meth,[Arg]})
* the Python binding provides useful tools for code generation
* seems to be the Future (tm) in the *nix world, and runs on Windows too.

  there's three parts to the gtknode. a main loop, some support
functions for object storage and marshalling, and a whole bunch of
generated wrapper functions.
  The main loop is a pretty generic c-node that simple-mindedly
receives lists of 2-tuples;

{atom('Func'), list(boolean()|integer()|float()|atom()|string())}

  it checks that there exists a function Func and calls it, passing a
pointer to the argument list. the Func's are wrapper functions
generated from the GTK header files, and unmarshalls and type checks
the arguments before the actual GTK functions are called.
  the return value of the GTK function is sent back to the Erlang node.

  a different kind of message is sent if there is an interesting event
in the GUI (e.g. a button is pressed), where "interesting" means specified
in the Glade file.

  REFERENCE

  the c-node is started thus;
gtknode node host regname cookie cnode-name otp-version

  when started, gtknode will connect to it's application by sending a
handshake message to {node@host, regname}.
  the messsage looks like this;
{{GtkPid,handshake}, []}

  the Erlang application sends messages to the gtknode using
GtkPid. messages look like this;

list({'Gtk_function', [Args]})

  E.g., if we have a GtkButton widget, named b1, and we want to use
these functions;

const gchar* gtk_button_get_label (GtkButton *button);
void         gtk_button_set_label (GtkButton *button, const gchar *label);

  we could send this;

GtkPid ! [{'Gtk_button_set_label',[b1,"foo"]},{'Gtk_button_get_label',[b1]}].

  and we would receive this;

{{GtkPid,reply}, [{ok,void},{ok,"foo"}]}

  signals are sent from gtknode if the signal handler for a specified
signal-widget combination is set to gn_sighandler. the signals look
like this;

{{GtkPid, signal}, {atom(WidgetName),atom(SignalName)}}

  E.g., if we delete the GtkWindow named window1 we'll get this signal

{{GtkPid, signal},{window1,'GDK_DELETE'}}

  given that we've requested it, of course.

  EXAMPLES

  the file src/gtknode.erl implements a controller/middleman for the
gtknode, it's quite instructive. it is recommended to use this instead of
 working directly against the c-node.
  the file examples/simple/simple.erl implements the Erlang side of a
GUI for a simple 'top' application. the GUI is specified in
examples/simple/simple.glade

  GETTING IT

http://github.com/massemanet/gtknode/

  STATUS

  stable since 2008



ChangeLog
commit ba102c4e5be06e3ea428fa475d05d7cce11d96fc
Author: mats cronqvist <masse@cronqvi.st>
Date:   Thu Dec 24 23:18:55 2015 +0100

    v1.0.4

src/gtknode.app.src

commit 7183d55fe968a8866ee64c4f8768c0237b192a7a
Author: mats cronqvist <masse@cronqvi.st>
Date:   Thu Dec 24 23:18:00 2015 +0100

    spelling

README.md

commit 7c4fcca0a021c7ae7f8632ded1c8ed3c6371ae1c
Author: Duncan McGreggor <oubiwann@gmail.com>
Date:   Sun Dec 20 19:46:04 2015 -0600

    Updated ignores.

.gitignore

commit bd19a341a2051bec55bfaa2f70799b34667a9de0
Author: Duncan McGreggor <oubiwann@gmail.com>
Date:   Sun Dec 20 18:06:54 2015 -0600

    Consistent use of lower-case.
    Also:
     * and formatting for var name
     * use standard colons

README.md

commit 657feef0cb9c114f40ce6625a4442d8fbe8600dc
Author: Duncan McGreggor <oubiwann@gmail.com>
Date:   Sun Dec 20 18:02:40 2015 -0600

    Added formating for another block of Erlang code.

README.md

commit 9b2c88039d4c33a37d2a9649a9c1d3502fdbb623
Author: Duncan McGreggor <oubiwann@gmail.com>
Date:   Sun Dec 20 17:54:44 2015 -0600

    Formatting tweaks

README.md

commit 2982aacb33de885795814b53ffa955cf0d1ab668
Author: Duncan McGreggor <oubiwann@gmail.com>
Date:   Sun Dec 20 17:52:18 2015 -0600

    Converted to Markdown

README
README.md

commit dcfc51c94b1253e3fe9a1f5dc554fc5a4e73da9c
Author: mats cronqvist <masse@cronqvi.st>
Date:   Tue Dec 1 15:27:41 2015 +0100

    fix readlink on mac

INSTALL
priv/examples/hello/Makefile
priv/examples/hello_world/Makefile
priv/examples/points/Makefile
priv/examples/top/Makefile
priv/examples/treeview/Makefile
priv/generator/generate.sh

commit 4d8a3edae04874a3dd5ab75d9b39353b6cb4b444
Author: mats cronqvist <masse@cronqvi.st>
Date:   Mon Nov 9 16:26:30 2015 +0100

    remove dialyzer no_spec flag

Makefile

commit 33706a3f7e73ada1796470137843bf0a02ec5877
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Nov 2 14:10:30 2015 +0000

    v1.0.2

ChangeLog
src/gtknode.app.src

commit 46462c1f2d55157f91cd82a7d6ef976994d402c4
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Nov 2 14:10:07 2015 +0000

    sh is old, use bash instead

bin/release.sh

commit 4602344ab6a8bdfab5fba0a4650ee267ba333b13
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Nov 2 14:09:07 2015 +0000

    v1.0.1

ChangeLog
src/gtknode.app.src

commit 36f76715d6d7d781ff63c39f199b92b95127e91f
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Nov 2 14:07:20 2015 +0000

    docs

INSTALL
README

commit e31a493bb61b7bec9f3b25f7179f5962a5607b8d
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Nov 2 11:51:31 2015 +0000

    make example tweak

priv/examples/hello/Makefile

commit 40fb2180f30869ab0670a900367465c2e9e11b6c
Author: mats cronqvist <masse@klarna.com>
Date:   Sat Oct 24 19:18:43 2015 +0000

    fix linker flags

priv/Makefile

commit 661cd60d7e01d77cb02b1468caef3f16bafdd017
Author: mats cronqvist <masse@klarna.com>
Date:   Wed Oct 14 13:27:10 2015 +0200

    v1.0.0

AUTHORS
ChangeLog
src/gtknode.app.src

commit 5944ab0d3a58421d87217e256539cfcd8f02140e
Author: mats cronqvist <masse@klarna.com>
Date:   Wed Oct 14 13:26:37 2015 +0200

    add release script

.gitignore
bin/release.sh

commit 32d496bc4dc1f2dad8a4069efbed783332bbcb51
Author: mats cronqvist <masse@klarna.com>
Date:   Wed Oct 14 13:16:35 2015 +0200

    quiet spurious dialyzer warning

src/gtknode.erl

commit 11dac181c06559f71974e710825d6fb603ea9a50
Author: mats cronqvist <masse@cronqvi.st>
Date:   Tue Oct 13 20:13:43 2015 +0200

    fix dialyzer make target

Makefile
src/gtknode.erl

commit 79bb10a8b379a4bbcf85dae07646a3de74f45a60
Author: mats cronqvist <masse@cronqvi.st>
Date:   Tue Oct 13 19:44:04 2015 +0200

    fix "make test" fail

Makefile

commit 39b55b6f28677d3b3d1f399323a1bcb00177f712
Author: mats cronqvist <masse@cronqvi.st>
Date:   Tue Oct 13 19:34:34 2015 +0200

    fix buggy makefile

priv/generator/Makefile

commit 83c632e49ee083d52ba832d90e95cab4ca6e2be2
Author: mats cronqvist <masse@klarna.com>
Date:   Tue Oct 13 16:53:32 2015 +0200

    more makefiling

priv/Makefile
priv/c_src/gtknode_gen.c
priv/generator/Makefile

commit f8781a25df8a1abf1186d062b941371f4a1910da
Author: mats cronqvist <masse@klarna.com>
Date:   Tue Oct 13 11:14:35 2015 +0200

    harmonize rebar/makefile

Makefile
priv/generator/generate.sh
rebar.config

commit d9cc9cce71a5b9b75abf5fd5f8a489558659daba
Author: mats cronqvist <masse@klarna.com>
Date:   Tue Oct 13 11:05:58 2015 +0200

    ebin dir has moved

priv/generator/generate.sh

commit ee7b80d2a5daaf05d3b0cea5926996656a258544
Author: mats cronqvist <masse@klarna.com>
Date:   Tue Oct 13 11:03:29 2015 +0200

    remove dead make target

priv/Makefile

commit c650a0e43269bd998aaff1e8e7204c3c16d506e7
Author: mats cronqvist <masse@klarna.com>
Date:   Tue Oct 13 11:02:30 2015 +0200

    find otp includes with escript

Makefile
priv/Makefile
priv/get_erl_root.es

commit d5d6dae41edcfcf474fa49d07d10007026bd43ac
Author: mats cronqvist <masse@klarna.com>
Date:   Tue Oct 13 10:07:26 2015 +0200

    reintroduce rebar

Makefile
c_src/gtknode.c
c_src/gtknode.h
c_src/gtknode_cnode.c
c_src/gtknode_gen.c
c_src/gtknode_glade.c
c_src/gtknode_internal.c
c_src/gtknode_marshal.c
c_src/gtknode_structs.c
priv/Makefile
priv/c_src/gtknode.c
priv/c_src/gtknode.h
priv/c_src/gtknode_cnode.c
priv/c_src/gtknode_gen.c
priv/c_src/gtknode_glade.c
priv/c_src/gtknode_internal.c
priv/c_src/gtknode_marshal.c
priv/c_src/gtknode_structs.c
priv/generator/generate.sh
rebar
rebar.config

commit e5c5bcc21b527f044290c1bf17ae640e2cc7f732
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Oct 12 15:54:47 2015 +0200

    fix examples

priv/examples/README
priv/examples/hello_world/hello_world.erl
priv/examples/points/points.erl
priv/examples/top/top.erl
priv/examples/treeview/treeview.erl

commit bf55565a95f37d14534b9f3d3b7034039fbea51b
Author: mats cronqvist <masse@klarna.com>
Date:   Mon Oct 12 15:32:39 2015 +0200

    makefiles

Makefile
priv/examples/Makefile

src/generator.erl
%%%-------------------------------------------------------------------
%%% File    : generate.erl
%%% Author  : Mats Cronqvist <qthmacr@mwux005>
%%% Description : generates C code that gtkNode_cnode can call.
%%%               each function marshals args and calls one gtk function.
%%%               reads the .defs files from h2def.py
%%%
%%% Created : 27 Oct 2004 by Mats Cronqvist <qthmacr@mwux005>
%%%-------------------------------------------------------------------
-module(generator).

-export([go/1]).
-import(filename,[dirname/1,basename/1]).
-import(lists,[foreach/2,foldl/3,reverse/1,sort/1]).

-record(func, {black=no,cname,pref,ret,const=no,paras=[]}).
-record(type, {what,cname,gtype,etype,cast}).

go([Root, Vsn, StructsFile | Files]) ->
  ets_new(types),
  ets_new(funcs),
  ets_new(bogus, [{keypos, 1}]),
  get_doc_links(Root),
  do_types_loop(Files), % populate the type table
  lists(),
  structs(StructsFile),
  basic_types(),
  R = generate_wrappers_loop(Vsn, Files),
  io:fwrite("~w ~p~n~n", [?MODULE, R]).

doc_header(FD)->
  io:fwrite(FD,"<html><head>"
            "<title>Functions wrapped by gtknode</title><body>"
            "<h2>Functions wrapped by gtknode</h2>",[]).
doc_footer(FD)->
  io:fwrite(FD,"</body></html>",[]).

get_doc_links("no")->
  %% don't put links in the html files
  ok;
get_doc_links("yes")->
  %% don't put links in the html files
  io:fwrite("configure flag --enable-gtkdoclinks takes an argument",[]);
get_doc_links(Root)->
  %% put in links to some place like these;
  %% "/usr/share/doc/libgtk2.0-doc",
  %% "http://developer.gnome.org/doc/API/2.0"
  F = fun(D) -> put_links(D,Root) end,
  try foreach(F, ["gtk","glib","gdk","gobject","gdk-pixbuf"])
  catch _:R -> io:fwrite("doc_links failed:~p~n",[R])
  end.

put_links(D,Root) ->
  Rt = string_join([Root,D,"ix01.html"],"/"),
  doc_links(Rt,get_ix(Rt)).

get_ix("http://"++_ = Doc) -> get_ix_http(Doc);
get_ix("file:/"++Doc) -> get_ix_file(Doc);
get_ix("/"++_ = Doc) -> get_ix_file(Doc).

get_ix_file(Doc) ->
  {ok,Bin} = file:read_file(Doc),
  binary_to_list(Bin).

get_ix_http(Doc) ->
  case os:getenv("http_proxy") of
    false -> ok;
    Proxy ->
      %% [http://][username:password@]proxyhost:port[/]
      case string:tokens(Proxy,"/")  of
        ["http:",Prx] -> ok;
        [Prx] -> ok
      end,
      case string:tokens(Prx,"@") of
        [_,Px] -> throw(proxy_authentication_not_supported);
        [Px] -> ok
      end,
      [H,P]  = string:tokens(Px,":"),
      httpc:set_options([{proxy,{{H,list_to_integer(P)},[]}}])
  end,
  ReqPid = spawn(fun() -> exit(httpc:request(Doc)) end),
  ReqMon = erlang:monitor(process,ReqPid),
  receive
    {'DOWN',ReqMon,_,_,{ok,{_,_,S}}} -> S;
    {'DOWN',ReqMon,_,_,I} -> throw({bad_http_reply,I})
  after
    9999 -> erlang:demonitor(ReqMon), exit(ReqPid,kill), throw(http_timeout)
  end.

doc_links(Root,S)->
  RE = "href=\"[A-Za-z-]+.html#id[0-9]+\">[a-z_]+ \\(\\)</a>",
  {match,Ms} = re:run(S,RE),
  io:fwrite("got ~p links for ~s~n",[length(Ms),basename(dirname(Root))]),
  foreach(fun({St,Le})-> do_doc_link(Root,string:substr(S,St,Le)) end, Ms).

do_doc_link(URL,S) ->
  case string:tokens(S,"=>< ,\"") of
    ["href",Href,"g"++Func,"()","/a"] ->
      put("G"++Func,"<a href=\""++dirname(URL)++"/"++Href++"\">");
    _ ->
      ok
  end.

do_types_loop([]) ->
  ok;
do_types_loop([_Prefix, FuncDefs, _FuncDefsWhiteList, _FuncDefsBlackList,
               _FuncWrappersCode, _WrappedFuncsList,
               _ErrorFuncsList, _ErrorTypesList| Tail]) ->
  do_types(FuncDefs),
  do_types_loop(Tail).

generate_wrappers_loop(Vsn, Files) ->
  gw_loop(Vsn, Files, []).

gw_loop(_Vsn, [], Results) -> Results;
gw_loop(Vsn, [Prefix, FuncDefs,
              FuncDefsWhiteList,FuncDefsBlackList,
              FuncWrappersCode, WrappedFuncsList,
              ErrorFuncsList,ErrorTypesList | Tail], Results) ->
  Res = generate_wrappers(Vsn, Prefix, FuncDefs,
                          FuncDefsWhiteList, FuncDefsBlackList,
                          FuncWrappersCode, WrappedFuncsList, ErrorFuncsList,
                          ErrorTypesList),
  gw_loop(Vsn, Tail, [Res | Results]).

generate_wrappers(Vsn, Prefix, _FuncDefs,
                  FuncDefsWhiteList, FuncDefsBlackList,
                  FuncWrappersCode, WrappedFuncsList,
                  ErrorFuncsList, ErrorTypesList) ->
  do_funcs(FuncDefsWhiteList), %populate the funcs table
  do_black(FuncDefsBlackList), %de-populate the funcs table
  {ok,FDc} = file:open(FuncWrappersCode, [write]),
  {ok,FDok} = file:open(WrappedFuncsList, [write]),
  {ok,FDcrap} = file:open(ErrorFuncsList, [write]),
  {ok,FDtypes} = file:open(ErrorTypesList, [write]),
  doc_header(FDok),
  vsn(Vsn, [FDc, FDok, FDcrap, FDtypes]),
  FRWAs = ets:match(funcs, {func, no, '$1', Prefix, '$2', '$3', '$4'}),
  foreach(fun(FRWA) -> gen_one(FDc, FDcrap, FDok, Prefix, FRWA) end, FRWAs),
  log_types(FDtypes),
  doc_footer(FDok),
  {Prefix, length(FRWAs), ets:lookup(bogus, good), ets:lookup(bogus, bad)}.

vsn(_, []) -> ok;
vsn(Vsn, [FD | FDs]) ->
  io:fwrite(FD, "/* GTK version: ~s */~n", [Vsn]),
  vsn(Vsn, FDs).

gen_one(FDc, FDcrap, FDok, Prefix, [Func,Ret,Const,Args]) ->
  unstick(),
  emit_head(Prefix, Func),
  emit_argdefs(Args),
  emit_retdef(Ret,Const),
  emit_ari_chk(Args),
  emit_arg_chk(Args),
  emit_call(Func, Ret, Args),
  emit_free(Args),
  emit_return(Ret),
  emit_final(),
  unstick(FDc,FDcrap,FDok).

emit_head(Prefix, Func) ->
  estick({UFunc = upcase_1st(Func), rem_pref(Func, Prefix)}),
  stick("/*******************************/~n"
        "gboolean ~s(int ARI, ei_x_buff *XBUF, char *B, int *I){~n~n", [UFunc]).

emit_argdefs([]) -> stick("~n",[]);
emit_argdefs([{Typ,Name,_Doc}|Args]) ->
  stick("  ~s ~s;~n", [Typ,Name]),
  emit_argdefs(Args).

emit_retdef(Ret,Const) ->
  case what_group(Ret) of
    object -> stick("  GObject* R;~n~n",[]);
    struct -> stick("  ~s R;~n~n", [Ret]);
    none -> stick("  /* no return value */~n~n",[]);
    crap -> stick("  ~s R;~n~n", ["CRAP_"++Ret]);
    list ->
      mark_as_crap("list_"++Ret),
      stick("  ~s R;~n~n", ["CRAP_list_"++Ret]);
    _ ->
      case Const of
        yes -> stick("  const ~s R; /* return value */~n~n", [Ret]);
        no -> stick("  ~s R; /* return value */~n~n", [Ret])
      end
  end.

emit_ari_chk(Args) ->
  stick("  if ( ! gn_check_arity(XBUF, ~p, ARI) )"
        " return FALSE;~n", [length(Args)]).

emit_arg_chk([]) -> ok;
emit_arg_chk([{Typ,Name,_Doc}|Args]) ->
  estick(Name),
  case Kind = what_group(Typ) of
    crap ->
      stick("  if ( ! gn_get_arg_CRAP(XBUF, B, I, ~p, &~s) )"
            " return FALSE;~n",
            [Typ,Name]);
    list ->
      stick("  if ( ! gn_get_arg_~s(XBUF, B, I, ~p, ~s&~s) )"
            " return FALSE;~n",
            [Kind,unstar(Typ),"(void**)",Name]);
    struct ->
      stick("  if ( ! gn_get_arg_~s(XBUF, B, I, ~p, ~s&~s) )"
            " return FALSE;~n",
            [Kind,unstar(Typ),"(void**)",Name]);
    basic ->
      stick("  if ( ! gn_get_arg_~s(XBUF, B, I, &~s) )"
            " return FALSE;~n",
            [unstar(Typ),Name]);
    flags ->
      stick("  if ( ! gn_get_arg_~s(XBUF, B, I, ~p, ~s&~s) )"
            " return FALSE;~n",
            [Kind,unstar(Typ),"(gint*)",Name]);
    enum ->
      stick("  if ( ! gn_get_arg_~s(XBUF, B, I, ~p, ~s&~s) )"
            " return FALSE;~n",
            [Kind,unstar(Typ),"(gint*)",Name]);
    object ->
      stick("  if ( ! gn_get_arg_~w(XBUF, B, I, ~s, ~s&~s) )"
            " return FALSE;~n",
            [Kind,what_gtype(unstar(Typ)),"(GObject**)",Name])
  end,
  emit_arg_chk(Args).

emit_call(Func, Ret, Args) ->
  case what_group(Ret) of
    none -> stick("  ~s(",[Func]);
    object -> stick("  R = (GObject*)~s(",[Func]);
    list -> stick("  R = (~s)~s(", ["CRAP_"++Ret,Func]);
    crap -> stick("  R = (~s)~s(", ["CRAP_"++Ret,Func]);
    _ -> stick("  R = ~s(", [Func])
  end,
  emit_call_arg(Args).

emit_call_arg([{_,Name,_}|As]) -> stick("~s", [Name]),emit_call_args(As);
emit_call_arg(A) -> emit_call_args(A).

emit_call_args([]) -> stick(");~n",[]);
emit_call_args([{_,Name,_}|Args]) ->
  stick(", ~s", [Name]),
  emit_call_args(Args).

emit_free([]) -> ok;
emit_free([{Type,Name,_}|Args]) ->
  case {Type,what_group(Type)} of
    {_,list} -> stick("  g_free(~s);~n", [Name]);
    {"gchar*",_}-> stick("  free(~s);~n", [Name]);
    {"char*",_}-> stick("  free(~s);~n", [Name]);
    _ -> ok
  end,
  emit_free(Args).

emit_return(Ret) ->
  case what_group(Ret) of
    basic ->
      {ET,Cast} = what_etype(Ret),
      stick("  gn_put_~w(XBUF,(~s)R);~n", [ET,Cast]);
    none -> stick("  gn_put_void(XBUF);~n", []);
    object -> stick("  gn_put_object(XBUF,R);~n", []);
    struct -> stick("  gn_put_struct(XBUF,~p,(void*)R);~n", [unstar(Ret)]);
    crap -> stick("  gn_put_CRAP(XBUF,~p,R);~n", [Ret]);
    Kind -> stick("  gn_put_~w(XBUF,~p,R);~n", [Kind, Ret])
  end.

emit_final() ->
  stick("  return TRUE;~n}~n",[]).

what_gtype(This) ->
  [#type{what=object,gtype=Type}] = ets:lookup(types,This),
  Type.

what_etype(Basic) ->
  [#type{what=basic,etype=Type,cast=Cast}] = ets:lookup(types,Basic),
  {Type,Cast}.

what_group([]) -> none;
what_group("none") -> none;
what_group(Typ) ->
  case {ets:lookup(types,Typ),ets:lookup(types,unstar(Typ))} of
    {[#type{what = object}],_} -> object;
    {[],[#type{what = object}]} -> object;
    {[#type{what = basic}],_} -> basic;
    {[#type{what = enum}],_} -> enum;
    {[#type{what = flags}],_} -> flags;
    {[#type{what = list}],_} -> list;
    {_,[#type{what = struct}]} -> struct;
    _ -> mark_as_crap(Typ)
  end.

mark_as_crap(Typ) ->
  put(crap,crap),
  ets_upd(bogus,{type,Typ}),
  crap.

estick(X) ->
  case get(efunc) of
    undefined -> put(efunc, {X,[]});
    {F,As} -> put(efunc,{F,[X|As]})
  end.

stick(Form, Args) ->
  S = io_lib:fwrite(Form, Args),
  case get(cfunc) of
    undefined -> put(cfunc, S);
    SS -> put(cfunc, SS++S)
  end.

unstick() -> erase(cfunc), erase(efunc), erase(crap).
unstick(FD1,FD2,FDok) ->
  case get(crap) of
    crap -> ets_upd(bogus,bad),cunstick(FD2);
    _ -> ets_upd(bogus,good),cunstick(FD1), eunstick(FDok)
  end,
  unstick().

cunstick(FD) -> io:fwrite(FD, "~s", [lists:flatten(get(cfunc))]).

eunstick(FD) ->
  {{Cname,_Ename},As} = get(efunc),
  case get(Cname) of
    undefined -> io:fwrite(FD, "<br>~s(",[Cname]);
    Link -> io:fwrite(FD, "<br>~s~s</a>(",[Link,Cname])
  end,
  eunstick_args(FD,As),
  io:fwrite(FD, ")~n",[]).

eunstick_args(_FD,[]) -> ok;
eunstick_args(FD,As) ->
  [A1|AT] = reverse(As),
  io:fwrite(FD,"~s",[upcase_1st(A1)]),
  foreach(fun(A)->io:fwrite(FD,",~s",[upcase_1st(A)]) end, AT).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
do_black(FN) ->
  bio:string(FN, fun do_black/2, nil).

do_black(Str,_State) ->
  case ets:lookup(funcs,Str) of
    [] -> ok;
    [Rec] -> ets:insert(funcs, Rec#func{black=yes})
  end.

do_funcs(FN) ->
  io:fwrite("~p~n", [FN]),
  bio:string(FN, fun do_funcl/2, nil).

do_funcl("",State) ->
  State;
do_funcl(";"++_,State) ->
  State;
do_funcl(")", bogus) ->
  nil;
do_funcl(")", {_,Data}) ->
  ets:insert(funcs,Data),
  nil;
do_funcl("(define-function "++_, nil) ->
  {func,#func{}};
do_funcl("(define-method "++_, nil) ->
  {meth,#func{}};
do_funcl("  (parameters", {Flag,Func}) ->
  {para, {Flag, Func}};
do_funcl(_, bogus) ->
  bogus;
do_funcl("  (varargs #t)", _) ->
  bogus;
do_funcl("  )", {para, {Flag, Func}}) ->
  {Flag, Func};
%%do_funcl("   )", {para, {Flag, Func}}) ->
%%    {Flag, Func};
do_funcl(Str, {para, {Flag, Func}}) ->
  OP = Func#func.paras,
  case string:tokens(Str,"'()\" ") of
    ["const-"++Para, Name|_Doc] ->
      {para, {Flag, Func#func{paras=OP++[{Para,Name,const}]}}};
    [Para, Name|_Doc] ->
      {para, {Flag, Func#func{paras=OP++[{Para,Name,no}]}}}
  end;
do_funcl("  (c-name \""++C, {Flag,Func}) ->
  {Flag,Func#func{cname=trnc(2,C), pref=pref(C)}};
do_funcl("  (return-type \""++C, {Flag,Func}) ->
  case trnc(2,C) of
    "const-"++R -> {Flag,Func#func{ret=R,const=yes}};
    R -> {Flag,Func#func{ret=R}}
  end;
do_funcl("  (of-object \""++C, {meth,Func}) ->
%%%    {meth,Func#func{this=trnc(2,C)}};
  {meth,Func#func{paras=[{trnc(2,C)++"*","object",no}]}};
do_funcl(_,State) ->
  State.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
lists() ->
  Lists = ["GType*"],
  foreach(fun ins_list/1, Lists).

ins_list(L) -> ets:insert(types,#type{what=list,cname=L,cast=""}).

structs(StructsFile) ->
  Structs = bio:string(StructsFile, fun do_structs/2, []),
  foreach(fun ins_struct/1, Structs).
do_structs(Str, Acc) ->
  case re:run(Str,"gn_construct_.*\\(") of
    nomatch -> Acc;
    {match,[{St,Le}]} -> [string:substr(Str,St+14,Le-14)|Acc]
  end.
ins_struct(S) -> ets:insert(types,#type{what=struct,cname=S,cast=""}).

basic_types() ->
  Basic = [{boolean, "int", ["gboolean","boolean"]},
           {string, "char*", ["gchar*","char*"]},
           {double, "double", ["gdouble","gfloat"]},
           {longlong, "long long", ["gint64","gint","int","size_t"]},
           {ulonglong, "unsigned long long",
            ["guint","guint8","guint16","guint32"]},
           {void,none,["void"]}],
  foreach(fun ins_basic/1, Basic).
ins_basic({ET,Cast,Bs}) ->
  foreach(fun(B) -> ins_basic(B,ET,Cast) end, Bs).
ins_basic(B,ET,Cast) ->
  ets:insert(types,#type{what=basic,cname=B,etype=ET,cast=Cast}).

do_types(FN) ->
  io:fwrite("~p~n", [FN]),
  bio:string(FN, fun do_typel/2, nil).

do_typel("",State) ->
  State;
do_typel(";"++_,State) ->
  State;
do_typel(")", {_,Data}) ->
  ets:insert(types,Data),
  nil;
do_typel("(define-boxed "++_, nil) ->
  {define_boxed,#type{what=boxed}};
do_typel("(define-enum "++_, nil) ->
  {define_enum,#type{what=enum}};
do_typel("(define-flags "++_, nil) ->
  {define_flags,#type{what=flags}};
do_typel("(define-object "++_, nil) ->
  {define_object,#type{what=object}};
do_typel("(define-pointer "++_, nil) ->
  {define_pointer,#type{what=pointer}};
do_typel("  (c-name \""++C, {Flag,Type}) ->
  {Flag,Type#type{cname=trnc(2,C)}};
do_typel("  (gtype-id \""++C, {Flag,Type}) ->
  {Flag,Type#type{gtype=trnc(2,C),cast=cast(trnc(2,C))}};
do_typel(_,State) ->
  State.

cast(C) ->
  [Prefix,"TYPE"|Toks] = string:tokens(C,"_"),
  string_join([Prefix|Toks], "_").

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
upcase_1st(Str) -> [hd(Str)+$A-$a|tl(Str)].

rem_pref([X|Func], [X|Prefix]) -> rem_pref(Func, Prefix);
rem_pref([$_|Func], []) -> Func.

pref(C) -> hd(string:tokens(C,"_")).

unstar(Str) ->
  case reverse(Str) of
    "*"++X -> reverse(X);
    _ -> Str
  end.

trnc(N,C) -> lists:sublist(C, length(C)-N).

string_join([Prefix|Toks], Sep) ->
  foldl(fun(Tok,O) -> O++Sep++Tok end, Prefix, Toks).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
log_types(FD) ->
  F = fun(E) -> io:fwrite(FD, "~4w - ~s~n", E) end,
  foreach(F,reverse(sort(ets:match(bogus,{{type,'$2'},'$1'})))).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
ets_new(Tab) ->
  catch ets:delete(Tab),
  ets_new(Tab, [{keypos,3}]).
ets_new(Tab, Opt) ->
  catch ets:delete(Tab),
  ets:new(Tab,[named_table,ordered_set]++Opt).
ets_upd(Tab, Key) ->
  case catch ets:update_counter(Tab, Key, 1) of
    {'EXIT',_} -> ets:insert(Tab, {Key,1});
    _ -> ok
  end.
